src/docker/idle_container_creator.rs

```diff
@@ -0,0 +1,138 @@
+use std::sync::mpsc::{sync_channel, Receiver};
+use std::thread;
+use std::time::Duration;
+
+use lazy_static::lazy_static;
+use parking_lot::Mutex;
+use subprocess::Popen;
+use tempfile::TempDir;
+
+use crate::docker::call_docker_async;
+use crate::error::{WorkerError, WorkerErrorKind};
+use crate::fs_utils::get_filename;
+use crate::named_pipe::NamedPipe;
+
+// NOTE: the number of idle containers on the system is CONTAINER_CACHE_CHANNEL_SIZE + CACHE_POPULATOR_COUNT
+const CONTAINER_CACHE_CHANNEL_SIZE: usize = 3;
+const CACHE_POPULATOR_COUNT: usize = 2;
```

Comment:
Can we talk about what these numbers mean?

Reply:
So `CONTAINER_CACHE_CHANNEL_SIZE` basically is how many containers can be cached in the channel between the cache populators and the user. The `CACHE_POPULATOR_COUNT` is how many threads are trying to populate that channel. Each thread may have started a container that isn't yet in the channel, so `CONTAINER_CACHE_CHANNEL_SIZE + CACHE_POPULATOR_COUNT = max number of cached containers`